src/H5Dmpio.c

```diff
@@ -2927,187 +3134,352 @@ H5D__obtain_mpio_mode(H5D_io_info_t *io_info, H5D_dset_io_info_t *di, uint8_t as
+
+            assert(di[dset_idx].dset->shared->ndims == di[dset_idx].dset->shared->layout.u.chunk.ndims - 1);
+            for (size_t dim_idx = 0; dim_idx < di[dset_idx].dset->shared->layout.u.chunk.ndims - 1; dim_idx++)
+                chunk_dims[dim_idx] = (hsize_t)di[dset_idx].dset->shared->layout.u.chunk.dim[dim_idx];
+
+            /* Get a dataspace for filling chunk memory buffers */
+            if (NULL == (curr_dset_info->fill_space = H5S_create_simple(
+                             di[dset_idx].dset->shared->layout.u.chunk.ndims - 1, chunk_dims, NULL)))
+                HGOTO_ERROR(H5E_DATASET, H5E_CANTINIT, FAIL, "unable to create chunk fill dataspace");
+
+            /* Initialize fill value buffer */
+            if (H5D__fill_init(&curr_dset_info->fb_info, NULL, (H5MM_allocate_t)H5D__chunk_mem_alloc,
+                               (void *)&di[dset_idx].dset->shared->dcpl_cache.pline,
+                               (H5MM_free_t)H5D__chunk_mem_free,
+                               (void *)&di[dset_idx].dset->shared->dcpl_cache.pline,
+                               &di[dset_idx].dset->shared->dcpl_cache.fill, di[dset_idx].dset->shared->type,
+                               di[dset_idx].dset->shared->type_id, 0, curr_dset_info->file_chunk_size) < 0)
+                HGOTO_ERROR(H5E_DATASET, H5E_CANTINIT, FAIL, "can't initialize fill value buffer");
+
+            curr_dset_info->fb_info_init = true;
+        }
+
+        /*
+         * If the dataset is incrementally allocated and hasn't been written
+         * to yet, the chunk index should be empty. In this case, a collective
+         * read of its chunks is essentially a no-op, so we can avoid that read
+         * later. If all datasets have empty chunk indices, we can skip the
+         * collective read entirely.
+         */
+        if (fill_msg->alloc_time == H5D_ALLOC_TIME_INCR)
+            if (H5D__chunk_index_empty(di[dset_idx].dset, &curr_dset_info->index_empty) < 0)
+                HGOTO_ERROR(H5E_DATASET, H5E_CANTGET, FAIL, "couldn't determine if chunk index is empty");
+
+        if ((fill_msg->alloc_time != H5D_ALLOC_TIME_INCR) || !curr_dset_info->index_empty)
+            chunk_list->all_dset_indices_empty = false;
+
+        if (curr_dset_info->chunk_idx_info.storage->ops->insert)
+            chunk_list->no_dset_index_insert_methods = false;
+
+        /*
+         * For multi-dataset I/O, use a hash table to keep a mapping between
+         * chunks and the cached info for the dataset that they're in. Otherwise,
+         * we can just use the info object directly if only one dataset is being
+         * worked on.
+         */
+        if (num_dset_infos > 1) {
+            HASH_ADD(hh, chunk_list->dset_info.dset_info_hash_table, dset_oloc_addr, sizeof(haddr_t),
+                     curr_dset_info);
+        }
+        else
+            chunk_list->dset_info.single_dset_info = curr_dset_info;
+        curr_dset_info = NULL;
```

Comment:
All of the above logic was moved from each sub-function creating the hash table each time into this single spot in the I/O setup. All the functions now index into the hash table according to the dataset object header address so that the same struct can be shared among all the sub-functions